### PR TITLE
Import of /learn/teaching-ocaml.md

### PR DESCRIPTION
**Pull request by yansh**
Following the discussion in #552, we reworked the teaching ocaml page with @amirmc

### CODE REVIEW

File: site/learn/teaching-ocaml.md [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+# Resources
+
+
+## Suggested Textbooks 
+
+
+This is a list of books and lecture notes that are being used for
+teaching OCaml.  There is a wider list of book to choose from at For
+the full list refer to
+[Textbook section](http://ocaml.org/learn/books.html)
```
Comment (amirmc) [inline site/learn/teaching-ocaml.md]: Broken sentence. Should read "There is also a wider list of books to choose from, which you can find on the [books page](http://ocaml.org/learn/books.html)"

File: site/learn/teaching-ocaml.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+1. [The OCaml System: Documentation and User's Manual](http://caml.inria.fr/pub/docs/manual-ocaml/) This the official User's Manual. It serves as a complete reference guide to OCaml. Updated for each version of OCaml, it contains the description of the language, of its extensions, and the documentation of the tools and libraries included in the official distribution. (Cornell)
+
```
Comment (amirmc) [inline site/learn/teaching-ocaml.md:81]: Why is "(Cornell)" at the end of the sentence?

File: site/learn/teaching-ocaml.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+1. [Lecture notes](http://www.seas.upenn.edu/~cis120/current/notes/120notes.pdf)
+   for Penn's CIS120.  Like the course, this book covers both OCaml
+   and Java, and is intended for students with no programming
+   background.
+1. [Real World OCaml](http://shop.oreilly.com/product/0636920024743.do),
+   also available [online](https://realworldocaml.org/).  Not really
+   intended as a textbook, and probably not ideal as a first programming book.  But it
+   has been used in a few university classes, including Harvard's CS51.
```
Comment (amirmc) [inline site/learn/teaching-ocaml.md]: "Not really intended as a textbook, and probably not ideal as a first programming book."

Why say this? It sounds unduly negative. Perhaps better to point out that Harvard's CS51 follows CS50 (so students have already been introduced to basic concepts).

File: site/learn/teaching-ocaml.md (hunk 2 above)
Comment (yansh) [inline site/learn/teaching-ocaml.md:81]: No reason. I removed it. Thanks.

Yan

File: site/learn/teaching-ocaml.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+
+
+## Teaching tools
+
+This section lists some of the available tools that are or can be used in a classroom or lab.
+
+### In the browser
+
+OCaml's bytecode is reletively starightforward to convert to Javascipt. This allows any OCaml code to run in the browser.
```
Comment (amirmc) [inline site/learn/teaching-ocaml.md]: Edit: "This allows any OCaml code to run in the browser, which is very convenient when delivering individual tutorials to a class as there's minimal set up time involved."

File: site/learn/teaching-ocaml.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+For a quick example of OCaml in the browser Run and explore OCaml without installing it.
```
Comment (amirmc) [inline site/learn/teaching-ocaml.md]: edit: "For a quick example of OCaml in the browser, try visiting any of the following."

File: site/learn/teaching-ocaml.md [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+1. [CSE 130: Programming Languages Course at UCSA](http://cseweb.ucsd.edu/classes/wi11/cse130/). Click on [Lectures] (http://cseweb.ucsd.edu/classes/wi11/cse130/) and [Discussion Notes] (http://cseweb.ucsd.edu/classes/wi11/cse130/). Have a look at Lec 7 (a pp presentation by Zach) for an inspiring introduction into map and fold.
+1. [CS 3110 Fall 2013 :: Data Structures and Functional Programming at Cornell](http://www.cs.cornell.edu/courses/cs3110/2013fa/lecture_notes.php). The lecturing materials explain OCaml on the basis of the fundamental concepts of functional programming languages.
+1. [CIS 500: Software Foundations at UPenn](http://www.seas.upenn.edu/~cis500/cis500-f05/index.html). A Software Fundations course with strong emphasis on Types. The [OCaml Tutorial] (http://www.seas.upenn.edu/~cis500/cis500-f05/resources/seas-ocaml.html) includes a gentle introduction to the OCaml interactive top-level system and compilation in OCaml.
+
+**Other sources**
+
+1. You can also check out the [OCaml tutorials section](http://ocaml.org/learn/tutorials/) for ideas and code.
```
Comment (amirmc) [inline site/learn/teaching-ocaml.md]: can also check out -> can check out

File: site/learn/teaching-ocaml.md (hunk 5 above)
Comment (yansh) [inline site/learn/teaching-ocaml.md]: Thanks @amirmc - just pushed the corrections.

File: site/learn/teaching-ocaml.md (hunk 6 above)
Comment (yansh) [inline site/learn/teaching-ocaml.md]: Thanks, corrected.

**Comment by amirmc** [comment]
@avsm @yallop @lpw25 @samoht and @yminsky.  As people who've taught OCaml, given tutorials etc, your input on this page would be appreciated.  I thought putting all the content on one page would make sense but we may wish to pare down the number of links so it's easier for a visitor to digest (e.g maybe pointing to one, nice IOCaml notebook might be preferable to many OCaml in the browser links).  

Please bear in mind that this page will evolve over time and although I hope we eventually converge on a recommended set of tools, I think we're not there yet.

Ashish, Christophe, I'll make a separate issue to break up this page using md.part so that we minimise duplication. It's easier to read and comment in its present form.

**Comment by yminsky** [comment]
Yeah, I think that page has too much.  There's also the OCaml-at-School
page, which I think has too little, but isn't (I think) too overwhelming.

https://github.com/ocaml/ocaml.org/wiki/OCaml-at-School

y

**Comment by amirmc** [comment]
I'd rather we only have the one page and that it be on OCaml.org than a wiki.

The content in this PR is an amalgamation of the OCaml at School page and the OCaml teaching resources page. 
What content do you think we can (should) drop?

**Comment by yallop** [comment]
It'd be more useful to have course names than course codes.  I'd also be inclined to put the course names before the universities.  For example
1. Programming Languages (University of Massachusetts Amherst)
2. Functional Programming (Princeton)
3. Programming Languages and Compilers (University of Illinois at Urbana-Champaign)

tells me much more about the courses and how they're likely to use OCaml than
1. University of Massachusetts Amherst (CMPSCI631)
2. Princeton (COS326)
3. University of Illinois at Urbana-Champaign (CS 421)

**Comment by yminsky** [comment]
Here are some things I'd nominate for removal:
- The troubleshooting section, which is as of now empty
- The "three fundamental concepts" section
- The "five of the most fundamental papers" section
- Some of the "tutorial" links, which I think are not suitable for classroom use.  In particular, the "beginner's guide to OCaml beginners guides", and the "OCaml Core" tutorials.  Maybe include the course-specific ones here, and just have a link to the more general list of tutorials on ocaml.org.
- Most of the "IDE" links.  I think that this may deserve its own page on ocaml.org, but I would restrict this page to things that are of focused interest to teachers.  Thus, if there was a good packaged IDE setup that was appropriate for student use, I'd link that from here.  But I wouldn't just have a collection of links on the various IDE solutions out there embedded in the page.
- I'd reorganize the "recommended textbooks" section.  I don't think we're really in a position to recommend textbooks, per se.  I think we should list textbooks that we know are used in courses, and then have a link to more general-purpose lists of textbooks on another page.

**Comment by yminsky** [comment]
Oh, and I agree that we should end up with something on ocaml.org.  I think the wiki was envisioned as just a way to get started.

I think that there should really be a focused ocaml-at-university page, different from the more general "Learn OCaml" info that's already on OCaml.org.  It does look like this page is shaping up to be that, which seems great to me.  Some of the material that I'd argue for removing from this page would sit happily elsewhere on ocaml.org.

**Comment by yansh** [comment]
On 07/11/14 15:41, yallop wrote:

> It'd be more useful to have course names than course codes. I'd also 
> be inclined to put the course names before the universities. For example
> 1. Programming Languages (University of Massachusetts Amherst)
> 2. Functional Programming (Princeton)
> 3. Programming Languages and Compilers (University of Illinois at
>    Urbana-Champaign)
> 
> tells me much more about the courses and how they're likely to use 
> OCaml than
> 1. University of Massachusetts Amherst (CMPSCI631)
> 2. Princeton (COS326)
> 3. University of Illinois at Urbana-Champaign (CS 421)
> 
> —
> 
> This makes sense, however I think the intial goal was  to list the 
> places where ocaml is taught. I think for early adopters it might be 
> important to know that unies like Cornell, Hardvard, Cambridge are using 
> ocaml.
> Perhaps we should change it to the following :

University of Massachusetts Amherst (Programming Language course)

Anyway, this is just an opinion, let me know what you guys prefer.

**Comment by yminsky** [comment]
I slightly prefer the universities first (for the reasons you mention), but
am happy with whatever outcome is chosen.  I like having course names, but
course numbers are important for people who want to look it up in course
catalogue.  We can maybe have all three --- university, course number,
course name?

y

**Comment by yminsky** [comment]
Also, it might be worth experimenting with organizing this as a table.

**Comment by pw374** [comment]
Just in case: http://philippewang.info/CL/?id=where_ocaml_is_used_for_teaching_purposes

**Comment by amirmc** [comment]
Thank you all for the comments. I concur with the point about course numbers so we should really keep them (everything seems to revolve around them in US schools).

@yansh Could you do the following?
- [ ] Reorder the course info as '{university} - {course_title} ({course_code})' -- Let's avoid a table for now as it's more work than anyone realises.
- [ ] Check if there are any additional courses on Philippe's page to incorporate
- [ ] In Tutorials and Exercises, remove the 'Other sources' (submit any new content to the existing tutorials page).
- [ ] Remove the IDE section - we should find elsewhere on the site for it (pls submit a separate PR for it).
- [ ] Ask @niksu about the Vagrant scripts he's been working on. They may belong in this page in the VM section.
- [ ] Add a note and link to the [new teaching mailing list](http://lists.ocaml.org/listinfo/teaching).

I think that covers everyone's comments and once complete, I'll be happy to merge this PR. I don't think we have any issue with duplicated content now either (which is nice).

**Comment by amirmc** [comment]
Just edited my previous comment to add an extra item about the new [teaching mailing list](http://lists.ocaml.org/listinfo/teaching).

**Comment by niksu** [comment]
The updated vagrant scripts have been upstreamed to the main repo at https://github.com/mirage/mirage-vagrant-vms

Two things to consider:
1. Currently the workflow assumes that the user is working on OSX. Adapting it for mainstream Linux distros shouldn't be hard. Not sure about Windows -- perhaps via cygwin?
2. As suggested by @avsm in a different thread, to make users' lives easier we could create a fork of https://github.com/mirage/mirage-vagrant-vms that contains provisioning scripts that install things like ocp-indent, merlin, perhaps utop, etc. Users then can generate usable programming environments using our scripts -- rather than having to discover and install the components they need. To start with, you'd need a list of components that you'd like installed in the resulting image, then we can work out how to get them installed and configured; I'd suggest that we start with something super simple.

**Comment by avsm** [comment]
Could also do the reverse: crete an ocaml/ocaml-vagrant-vms which has all the OCaml development environment setup, and then base the mirage/mirage-vagrant-vms off this base image to add the Xen hypervisor and mirage libraries as a fairly trivial script on the base.

-a

**Comment by niksu** [comment]
I guess the kind of people who might want to play with mirage-vagrant-vms wouldn't mind having extra OCaml-related tools in their image.
In any case, we could set up the relation between to repos neatly using git submodules perhaps.

**Comment by niksu** [comment]
Nah on second thoughts submodules don't seem suitable here -- a simple fork seems good enough.

**Pull request merged**